[Core] Extending FindIntersectedGeometricalObjectsProcess to any entity

loumalouomega opened the pull request:
Fixes #4706 part of  #4263. This allows to use FindIntersectedGeometricalObjectsProcess withElements and conditions

loumalouomega committed:
Extending FindIntersectedGeometricalObjectsProcess to any entity

loumalouomega committed:
Update test

loumalouomega committed:
Update calculate_distance_to_skin_process

loumalouomega committed:
Update calculate_embedded_nodal_variable_from_skin_process

loumalouomega committed:
Update calcualte_discontinous_distance_to_skin_process

loumalouomega committed:
Exposing to python

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -25,231 +27,457 @@
+FindIntersectedGeometricalObjectsProcess<TEntity>::FindIntersectedGeometricalObjectsProcess(
+    Model& rModel,
+    Parameters ThisParameters
+    ) : mrModelPart1(rModel.GetModelPart(ThisParameters["first_model_part_name"].GetString())),
```
Maybe this is a good chance to use better names for the model parts. Even though the member variables keep their current names, I'd rename the Parameters constructor arguments to something more self-explaining like `intersected_model_part_name` and `intersecting_model_part_name`.

loumalouomega replied:
I prefer to do this in a different PR, as I said, this is part #4263 and I want to avoid conflicts

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -25,231 +27,457 @@
+    Parameters ThisParameters
+    ) : mrModelPart1(rModel.GetModelPart(ThisParameters["first_model_part_name"].GetString())),
+        mrModelPart2(rModel.GetModelPart(ThisParameters["second_model_part_name"].GetString())),
+        mThisParameters(ThisParameters)
```
Why do we need to save the Parameters as a member variable?

loumalouomega replied:
For the sake of generality (it makes more sense in #4263)

rubenzorrilla replied:
I wouldn't save them if there is no necessity.

loumalouomega replied:
This is removed, look the outdated label

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -25,231 +27,457 @@
+
+    IndexType counter = 0;
+    rResults.resize(number_of_conditions);
+    for (auto& r_condition : r_conditions_array) {
```
For the sake of avoiding code duplication, I'd create a method `GetIntersectingEntities` which is the one specialized for `Element`and `Condition`. Thus,  the loop becomes something like `for (auto &r_entity : this->GetIntersectingEntities())`. In this way it is not needed to specialize the `FindIntersectedSkinObjects` method.
(we should check which one is the intersected and intersecting one, it wouldn't be the first time that I switch them :sweat_smile: )

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -25,231 +27,457 @@
+    const auto it_elem_begin = mrModelPart1.ElementsBegin();
+
+    #pragma omp parallel for private(leaves)
+    for (int i = 0; i < number_of_elements; i++) {
```
I think that here we can apply the suggestion above as well.

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -25,231 +27,457 @@
+template<>
+std::size_t FindIntersectedGeometricalObjectsProcess<Element>::WorkingSpaceDimension()
+{
+    return mrModelPart1.Elements().begin()->GetGeometry().WorkingSpaceDimension();
```
And here as well.

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -25,231 +27,457 @@
-            mOctree.Insert(*(i_element).base());
-        }
+    // Iterate over the elements
+    for (auto it_elem = mrModelPart2.ElementsBegin(); it_elem != mrModelPart2.ElementsEnd(); it_elem++) {
```
Here we do the same but for the other model part, so we can create an specialization for each one of the model parts.

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.h:
```diff
@@ -7,355 +7,667 @@
-        bool HasIntersection3D(Element::GeometryType& rFirstGeometry, Element::GeometryType& rSecondGeometry);
-        void FindIntersectedSkinObjects(Element& rElement1, std::vector<OctreeType::cell_type*>& leaves, PointerVector<GeometricalObject>& rResults);
+        /// Assignment operator.
+        DistanceSpatialContainersConfigure& operator=(DistanceSpatialContainersConfigure const& rOther);
```
```suggestion
        DistanceSpatialContainersConfigure& operator=(DistanceSpatialContainersConfigure const& rOther) = delete;
```

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.h:
```diff
@@ -7,355 +7,667 @@
+        DistanceSpatialContainersConfigure& operator=(DistanceSpatialContainersConfigure const& rOther);
 
+        /// Copy constructor.
+        DistanceSpatialContainersConfigure(DistanceSpatialContainersConfigure const& rOther);
```
```suggestion
        DistanceSpatialContainersConfigure(DistanceSpatialContainersConfigure const& rOther) = delete;
```

rubenzorrilla replied:
Is there any reason to avoid deleting the copy constructor?

loumalouomega replied:
I updated manually

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.h:
```diff
@@ -7,355 +7,667 @@
+    FindIntersectedGeometricalObjectsProcess(FindIntersectedGeometricalObjectsProcess const& rOther) = delete;
+
+    /// Destructor.
+    ~FindIntersectedGeometricalObjectsProcess() override {}
```
```suggestion
    ~FindIntersectedGeometricalObjectsProcess() override = default;
```

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.h:
```diff
@@ -7,355 +7,667 @@
+    ModelPart& mrModelPart1;          /// First model part
+    ModelPart& mrModelPart2;          /// Second model part
+    OctreeType mOctree;               /// The octree structucture that performs the search
+    Parameters mThisParameters;       /// The configuration parameters
```
Why do we need to save them?

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.h:
```diff
@@ -7,355 +7,667 @@
+    ///@{
+
+    /// Assignment operator.
+    FindIntersectedGeometricalObjectsProcess& operator=(FindIntersectedGeometricalObjectsProcess const& rOther);
```
```suggestion
    FindIntersectedGeometricalObjectsProcess& operator=(FindIntersectedGeometricalObjectsProcess const& rOther) = delete;
```
And move it to the life cycle section

rubenzorrilla replied:
Any reason to avoid deleting it?

loumalouomega replied:
I added manually, don't trust GitHub

rubenzorrilla commented on kratos/python/add_processes_to_python.cpp:
```diff
@@ -539,7 +539,17 @@ void  AddProcessesToPython(pybind11::module& m)
+    .def(py::init<ModelPart&,ModelPart&, Parameters>())
+    ;
+
+    py::class_<FindIntersectedGeometricalObjectsProcess<Element>, FindIntersectedGeometricalObjectsProcess<Element>::Pointer, Process>(m,"FindIntersectedElementsProcess")
```
```suggestion
    py::class_<FindIntersectedGeometricalObjectsProcess<Element>, FindIntersectedGeometricalObjectsProcess<Element>::Pointer, Process>(m,"ElementBasedFindIntersectedGeometricalObjectsProcess")
```

rubenzorrilla replied:
Is there any place in where we directly use this in Python? If there isn't I wouldn't export it.

loumalouomega committed:
Deleting assigment operator

rubenzorrilla commented on kratos/python/add_processes_to_python.cpp:
```diff
@@ -539,7 +539,17 @@ void  AddProcessesToPython(pybind11::module& m)
+    .def(py::init<Model&, Parameters>())
+    ;
+
+    py::class_<FindIntersectedGeometricalObjectsProcess<Condition>, FindIntersectedGeometricalObjectsProcess<Condition>::Pointer, Process>(m,"FindIntersectedConditionsProcess")
```
```suggestion
    py::class_<FindIntersectedGeometricalObjectsProcess<Condition>, FindIntersectedGeometricalObjectsProcess<Condition>::Pointer, Process>(m,"ConditionBasedFindIntersectedGeometricalObjectsProcess")
```

rubenzorrilla reviewed (changes requested):
In general terms, I agree with the implementation. There are however some improvements that can be done to avoid code duplication.

(aesthetic changes again)

loumalouomega committed:
Default destructor

loumalouomega committed:
Removing Parameters

loumalouomega committed:
Reducing code duplication

loumalouomega committed:
Update name of model parts

loumalouomega committed:
More code reduction

loumalouomega commented:
Comments addressed @rubenzorrilla

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -462,13 +409,31 @@ void FindIntersectedGeometricalObjectsProcess<TEntity>::FindIntersectedSkinObjec
+template<>
+PointerVectorSet<Element, IndexedObject>::ContainerType& FindIntersectedGeometricalObjectsProcess<Element>::GetIntersectingEntities()
+{
+    return mrModelPartIntersected.ElementsArray();
```
Are you sure of this? This is supposed to be the skin (intersecting model part). In here you are taking the elements, or conditions in the method below, from the volume part (intersected model part). If I'm not remembering wrong, the volume part must always be defined in terms of elements.

loumalouomega replied:
I just changed the name according to you suggestion :P

rubenzorrilla replied:
Indeed the method is named `GetIntersectingElements` and you are working with the `mrModelPartIntersected`.

rubenzorrilla replied:
> I just changed the name according to you suggestion :P

Well, it is not a suggestion. I guess the distance algorithms wouldn't work with the current implementation.

rubenzorrilla replied:
This has to return the intersecting entities, which are the ones that can be defined as elements (as they are now) or as conditions (what we suggested to implement). The intersected model part will never be defined as a condition.

loumalouomega replied:
Well I just mistake the names..., changing

rubenzorrilla replied:
> Well I just mistake the names..., changing

Were the tests working? If they were we need to further improve them...

loumalouomega replied:
Yes, I only changed the names, nothing else. Just the names are wrong

pooyan-dadvand replied:
Wouldn't be better to add a `GetEntities` method to the ModelPart? There are other issues with the same request and we can finish all these kind of implementations much easier and cleaner!

loumalouomega replied:
Done

loumalouomega committed:
Updating name for base class

loumalouomega committed:
Revert "Updating name for base class"

This reverts commit 531a6f61896c64c0d24ac30c1f8b0b77a1a711dd.

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -28,8 +28,8 @@ template<class TEntity>
 FindIntersectedGeometricalObjectsProcess<TEntity>::FindIntersectedGeometricalObjectsProcess(
     ModelPart& rPart1,
     ModelPart& rPart2
-    ) : mrModelPartIntersected(rPart1),
```
Part 1 is supposed to be the volume part, which is the intersected one. This was correct. What was wrong was the new methods you created.

loumalouomega replied:
The methods I created followed exactly was was done until now

rubenzorrilla replied:
It is not a matter of search and replace, the implementation of the new specialized methods was wrong.

loumalouomega replied:
https://github.com/KratosMultiphysics/Kratos/blob/4e433cb22a51ecc2a8ea444adbeb4b0bedda31c6/kratos/processes/find_intersected_geometrical_objects_process.cpp#L41, in that case waht was wrong was  the name of the method

loumalouomega replied:
Fixed

loumalouomega committed:
Correct name

rubenzorrilla commented:
I'm compiling the branch to push the correct changes.

loumalouomega commented:
> I'm compiling the branch to push the correct changes.

Please no, I have 3 PR depending of this

loumalouomega committed:
Fix

rubenzorrilla commented:
> > I'm compiling the branch to push the correct changes.
> 
> Please no, I have 3 PR depending of this

I'm sorry but after multiple trials it is not correct yet. This is an extremely important utility and I don't want to mess it up. You can cherry-pick my commit in each one of your 3 three depending branches.

loumalouomega commented:
I updated the name, what is exactly wrong?

loumalouomega commented:
The name of the method was suggested by you

rubenzorrilla commented:
I don't care the name of the method, the problem is that you are retrieving the containers from the wrong model part. Besides now it is much difficult to follow since the names of the member model parts have changed (I only asked to change the arguments in the new json constructor).

loumalouomega commented:
> I don't care the name of the method, the problem is that you are retrieving the containers from the wrong model part. Besides now it is much difficult to follow since the names of the member model parts have changed (I only asked to change the arguments in the new json constructor).

I haven't changed the model parts. It is exactly the same model part as it was before. Previously the model part was the mrModelPart1:

https://github.com/KratosMultiphysics/Kratos/blob/4e433cb22a51ecc2a8ea444adbeb4b0bedda31c6/kratos/processes/find_intersected_geometrical_objects_process.cpp#L41

I maintained that at it was, I changed the name according to your suggestions. This is working exactly the same. The tests are passing, but not only the existing tests, but the 12 tests I added in the other PR

rubenzorrilla commented:
I don't know what happened but I wasn't looking the last version of the files. I will review the current status of the branch in my machine.

loumalouomega commented:
> I don't know what happened but I wasn't looking the last version of the files. I will review the current status of the branch in my machine.

Never trust github, it is always including the not updated version

rubenzorrilla commented:
OK, I've just checked the last version. I think that the implementation is now coincident with the previous one. However, this would allow to consider volume meshes defined as conditions but, will it work when the skin (intersecting model part) is defined with conditions as requested in #4706?

loumalouomega commented:
If you mean if will work when combining conditions/elements.Not now, it can be done

loumalouomega commented:
I am first solving the issue with appveyor

rubenzorrilla commented:
> If you mean if will work when combining conditions/elements.Not now, it can be done

I'm afraid that this might imply to modify the Octree...

loumalouomega commented:
> > If you mean if will work when combining conditions/elements.Not now, it can be done
> 
> I'm afraid that this might imply to modify the Octree...

Nope, not necessary :D

loumalouomega committed:
Windows compilation

loumalouomega committed:
Merge branch 'master' into core/FindIntersectedGeometricalObjectsProcess-entities

loumalouomega committed:
Linux compilation issue

loumalouomega commented:
> > > If you mean if will work when combining conditions/elements.Not now, it can be done
> > 
> > 
> > I'm afraid that this might imply to modify the Octree...
> 
> Nope, not necessary :D

Well 4 lines (not much :P)

loumalouomega committed:
Adding template argument to octree_binary

loumalouomega committed:
Generalizing combinations

@rubenzorrilla requested

loumalouomega committed:
Adding test

loumalouomega commented:
Done @rubenzorrilla

loumalouomega committed:
Expose to python (I)

loumalouomega committed:
Removing

rubenzorrilla commented on kratos/python/add_processes_to_python.cpp:
```diff
@@ -552,6 +552,23 @@ void  AddProcessesToPython(pybind11::module& m)
     .def(py::init<Model&, Parameters>())
     ;
 
+    py::class_<FindIntersectedGeometricalObjectsProcess<Element, Condition>, FindIntersectedGeometricalObjectsProcess<Element, Condition>::Pointer, Process>(m,"FindIntersectedElementsConditionsProcess")
```
As I told you before, I wouldn't export this to Python since it is not needed in any place. Besides, it makes no sense at all to call this process from Python, since the it only sets the Octree and compute the intersections accordingly, which are useless if called as a standalone utility.

loumalouomega commented:
I don't think so, It sets the flag selected. In any case I will remove.

El mié., 24 abr. 2019 9:50, Rubén Zorrilla <notifications@github.com>
escribió:

> *@rubenzorrilla* commented on this pull request.
> ------------------------------
>
> In kratos/python/add_processes_to_python.cpp
> <https://github.com/KratosMultiphysics/Kratos/pull/4707#discussion_r277994749>
> :
>
> > @@ -552,6 +552,23 @@ void  AddProcessesToPython(pybind11::module& m)
>      .def(py::init<Model&, Parameters>())
>      ;
>
> +    py::class_<FindIntersectedGeometricalObjectsProcess<Element, Condition>, FindIntersectedGeometricalObjectsProcess<Element, Condition>::Pointer, Process>(m,"FindIntersectedElementsConditionsProcess")
>
> As I told you before, I wouldn't export this to Python since it is not
> needed in any place. Besides, it makes no sense at all to call this process
> from Python, since the it only sets the Octree and compute the
> intersections accordingly, which are useless if called as a standalone
> utility.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/4707#pullrequestreview-229947463>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEYQZAFFVYN45TDHS6JBYTTPSAGMPANCNFSM4HHMEZ6Q>
> .
>

loumalouomega committed:
Reverting python exposition

loumalouomega commented:
Done @rubenzorrilla

loumalouomega committed:
Fix

rubenzorrilla commented:
So far I agree with the current state of the implementation (just have a look to my comments about the copy constructors). In any case, I would like @pooyan-dadvand to have an overview to this. Thanks for the effort.

rubenzorrilla committed:
Delete

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

rubenzorrilla committed:
Delete

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

loumalouomega committed:
Merge branch 'master' into core/FindIntersectedGeometricalObjectsProcess-entities

loumalouomega committed:
Adding entities utilities

loumalouomega committed:
Reducing  explicit templates

loumalouomega commented:
Suggestions adressed

pooyan-dadvand reviewed (commented):
It is an improvement to the previous ones but is over engineered with too many templates. Consider that most of the methods except the ones for getting the entity from modelpart can be implemented without templates and taking `GeometricalObject` as argument

loumalouomega commented:
> It is an improvement to the previous ones but is over engineered with too many templates. Consider that most of the methods except the ones for getting the entity from modelpart can be implemented without templates and taking `GeometricalObject` as argument

What do you suggest?, to remove the template arguments and passing the entities type via parameters?

pooyan-dadvand commented:
if you can use `GeometricalObject` instead of template typed argument, it would be better to use the `GeometricalObject`

loumalouomega commented:
> if you can use `GeometricalObject` instead of template typed argument, it would be better to use the `GeometricalObject`

OK

loumalouomega committed:
Merge branch 'master' into core/FindIntersectedGeometricalObjectsProcess-entities

loumalouomega committed:
Reverting specifications

loumalouomega committed:
Reverting changes

loumalouomega committed:
Refactoring without templates (I)

loumalouomega committed:
Merge branch 'master' into core/FindIntersectedGeometricalObjectsProcess-entities

loumalouomega committed:
Update test

loumalouomega committed:
Missing override

loumalouomega committed:
Adding direct intersection (I)

loumalouomega commented:
I adressed all comments

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -347,11 +347,20 @@ bool FindIntersectedGeometricalObjectsProcess::HasIntersection(
+        if (local_space_dimension == 2) {
+            return this->HasIntersection2D(rFirstGeometry, rSecondGeometry);
+        } else {
+            return this->HasDirectIntersection2D(rFirstGeometry, rSecondGeometry);
```
The difference of this new method with respect to the current `HasIntersection2D` is that it doesn't check that the intersecting geometry is inside the intersected candidate one. Am I wrong?

If this is the case, can you describe a use case in which it is required?

pooyan-dadvand replied:
As @rubenzorrilla asked, what is the difference?

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -347,11 +347,20 @@ bool FindIntersectedGeometricalObjectsProcess::HasIntersection(
+        if (local_space_dimension == 3) {
+            return this->HasIntersection3D(rFirstGeometry, rSecondGeometry);
+        } else {
+            return this->HasDirectIntersection3D(rFirstGeometry, rSecondGeometry);
```
Same

loumalouomega commented:
It is not needed, because the geometries are in a lower level, the nodes
cannot been inside without intersection.

El dom., 28 abr. 2019 2:25, Rubén Zorrilla <notifications@github.com>
escribió:

> *@rubenzorrilla* commented on this pull request.
> ------------------------------
>
> In kratos/processes/find_intersected_geometrical_objects_process.cpp
> <https://github.com/KratosMultiphysics/Kratos/pull/4707#discussion_r279170211>
> :
>
> > @@ -347,11 +347,20 @@ bool FindIntersectedGeometricalObjectsProcess::HasIntersection(
>      GeometryType& rSecondGeometry
>      )
>  {
> -    const IndexType work_dim = rFirstGeometry.WorkingSpaceDimension(); // TODO: DOMAIN_SIZE should be considered for consistency with other implementations
> -    if (work_dim == 2) {
> -        return this->HasIntersection2D(rFirstGeometry, rSecondGeometry);
> +    const IndexType working_space_dimension = rFirstGeometry.WorkingSpaceDimension(); // TODO: DOMAIN_SIZE should be considered for consistency with other implementations
> +    const IndexType local_space_dimension = rFirstGeometry.LocalSpaceDimension();
> +    if (working_space_dimension == 2) {
> +        if (local_space_dimension == 2) {
> +            return this->HasIntersection2D(rFirstGeometry, rSecondGeometry);
> +        } else {
> +            return this->HasDirectIntersection2D(rFirstGeometry, rSecondGeometry);
>
> The difference of this new method with respect to the current
> HasIntersection2D is that it doesn't check that the intersecting geometry
> is inside the intersected candidate one. Am I wrong?
>
> If this is the case, can you describe a use case in which it is required?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/4707#pullrequestreview-231439977>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEYQZAEEY4QHUZDYKZDEJYLPSTG3VANCNFSM4HHMEZ6Q>
> .
>

loumalouomega commented:
Intersection between lines in 2D or triangles / quadrilaterals in 3D for
example.

El dom., 28 abr. 2019 2:31, Vicente Mataix Ferrándiz <
vicente.mataix@polytechnique.edu> escribió:

> It is not needed, because the geometries are in a lower level, the nodes
> cannot been inside without intersection.
>
> El dom., 28 abr. 2019 2:25, Rubén Zorrilla <notifications@github.com>
> escribió:
>
>> *@rubenzorrilla* commented on this pull request.
>> ------------------------------
>>
>> In kratos/processes/find_intersected_geometrical_objects_process.cpp
>> <https://github.com/KratosMultiphysics/Kratos/pull/4707#discussion_r279170211>
>> :
>>
>> > @@ -347,11 +347,20 @@ bool FindIntersectedGeometricalObjectsProcess::HasIntersection(
>>      GeometryType& rSecondGeometry
>>      )
>>  {
>> -    const IndexType work_dim = rFirstGeometry.WorkingSpaceDimension(); // TODO: DOMAIN_SIZE should be considered for consistency with other implementations
>> -    if (work_dim == 2) {
>> -        return this->HasIntersection2D(rFirstGeometry, rSecondGeometry);
>> +    const IndexType working_space_dimension = rFirstGeometry.WorkingSpaceDimension(); // TODO: DOMAIN_SIZE should be considered for consistency with other implementations
>> +    const IndexType local_space_dimension = rFirstGeometry.LocalSpaceDimension();
>> +    if (working_space_dimension == 2) {
>> +        if (local_space_dimension == 2) {
>> +            return this->HasIntersection2D(rFirstGeometry, rSecondGeometry);
>> +        } else {
>> +            return this->HasDirectIntersection2D(rFirstGeometry, rSecondGeometry);
>>
>> The difference of this new method with respect to the current
>> HasIntersection2D is that it doesn't check that the intersecting
>> geometry is inside the intersected candidate one. Am I wrong?
>>
>> If this is the case, can you describe a use case in which it is required?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/KratosMultiphysics/Kratos/pull/4707#pullrequestreview-231439977>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AEYQZAEEY4QHUZDYKZDEJYLPSTG3VANCNFSM4HHMEZ6Q>
>> .
>>
>

rubenzorrilla commented:
Can you better explain?

What I see is that right now, if the intersecting entity (e.g. 2D line) completely lies inside the intersected entity (e.g. 2D triangle), meaning that no edge of the triangle intersects the line,  this was considered as intersection in any case. Your new function doesn't consider this as intersection.

If we move to the line line scenario, this geometry inside another only makes sense for the "one line over the other" case, which is already considered by the `IntersectionUtilities`. If we move to the 3D case, for instance a 3D triangle intersecting a 3D quadrilateral, the intersecting triangle may completely lie inside the intersected quadrilateral, which with the current implementation will be an intersection but no with your new function.

Is this the case?

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -8,248 +8,476 @@
+        auto it_cond = it_conditions_begin + i;
+        leaves.clear();
+        IdentifyNearEntitiesAndCheckEntityForIntersection(*(it_cond.base()), leaves);
     }
```
Now you create the Octree for both the elements and conditions of the intersected model part? Does this work if such intersected model part has both types of entities? If it does, doesn't this introduce an overhead? Maybe @pooyan-dadvand  can better clarify this.

loumalouomega replied:
You cannot if you want to be able to assign flags, because the GeometricalObject does not have flags

loumalouomega replied:
And the idea of the process is to assign the flag SELECTED

rubenzorrilla replied:
You mean that now it is needed to create the leaves of both elements and conditions because the `GeometricalObject` can't handle flags. Is this true?

rubenzorrilla commented on kratos/processes/find_intersected_geometrical_objects_process.cpp:
```diff
@@ -8,248 +8,476 @@
+
+std::size_t FindIntersectedGeometricalObjectsProcess::WorkingSpaceDimension()
+{
+    auto& r_elements_array = mrModelPartIntersected.Elements();
```
```suggestion
    const auto& r_elements_array = mrModelPartIntersected.Elements();
```
(just suggesting if you didn't try it)

rubenzorrilla committed:
Missing const

Co-Authored-By: loumalouomega <vicente.mataix@polytechnique.edu>

loumalouomega commented:
> Can you better explain?
> 
> What I see is that right now, if the intersecting entity (e.g. 2D line) completely lies inside the intersected entity (e.g. 2D triangle), meaning that no edge of the triangle intersects the line, this was considered as intersection in any case. Your new function doesn't consider this as intersection.
> 
> If we move to the line line scenario, this geometry inside another only makes sense for the "one line over the other" case, which is already considered by the `IntersectionUtilities`. If we move to the 3D case, for instance a 3D triangle intersecting a 3D quadrilateral, the intersecting triangle may completely lie inside the intersected quadrilateral, which with the current implementation will be an intersection but no with your new function.
> 
> Is this the case?

I understand your  point

loumalouomega committed:
@rubenzorrilla comment

loumalouomega commented:
@rubenzorrilla I addressed your comment

loumalouomega commented:
Yes, is this true, is the only reason I did it. If we move the flag from
elements and conditions to geometrical object it can be avoided.

El dom., 28 abr. 2019 17:33, Rubén Zorrilla <notifications@github.com>
escribió:

> *@rubenzorrilla* commented on this pull request.
> ------------------------------
>
> In kratos/processes/find_intersected_geometrical_objects_process.cpp
> <https://github.com/KratosMultiphysics/Kratos/pull/4707#discussion_r279199476>
> :
>
> >      }
>
> -    /// Print information about this object.
> -    void FindIntersectedGeometricalObjectsProcess::PrintInfo(std::ostream& rOStream) const {
> -        rOStream << Info();
> +    // Iterate over conditions
> +    auto& r_conditions_array = mrModelPartIntersected.Conditions();
> +    const SizeType number_of_conditions = r_conditions_array.size();
> +
> +    const auto it_conditions_begin = r_conditions_array.begin();
> +
> +    #pragma omp parallel for private(leaves)
> +    for (int i = 0; i < static_cast<int>(number_of_conditions); i++) {
> +        auto it_cond = it_conditions_begin + i;
> +        leaves.clear();
> +        IdentifyNearEntitiesAndCheckEntityForIntersection(*(it_cond.base()), leaves);
>      }
>
> You mean that now it is needed to create the leaves of both elements and
> conditions because the GeometricalObject can't handle flags. Is this true?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/4707#discussion_r279199476>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEYQZAFHJEHDZ5QWOFPELGLPSW6QLANCNFSM4HHMEZ6Q>
> .
>

rubenzorrilla commented:
> Yes, is this true, is the only reason I did it. If we move the flag from elements and conditions to geometrical object it can be avoided. 

And are you sure that this doesn't give any problem? @pooyan-dadvand

rubenzorrilla commented:
> @rubenzorrilla I addressed your comment

So now which is the difference between the original method and the one you added? I think that they are exactly the same.

loumalouomega commented:
> > @rubenzorrilla I addressed your comment
> 
> So now which is the difference between the original method and the one you added? I think that they are exactly the same.

It does not iterate over the edges/faces. It computes directly with the geometry. This is because @RiccardoRossi says that the discussion from #4221 and #4222 will take long

loumalouomega commented:
> > Yes, is this true, is the only reason I did it. If we move the flag from elements and conditions to geometrical object it can be avoided.
> 
> And are you sure that this doesn't give any problem? @pooyan-dadvand

No, and is something I should not decide, because is a design thing

rubenzorrilla commented:
> > > @rubenzorrilla I addressed your comment
> > 
> > 
> > So now which is the difference between the original method and the one you added? I think that they are exactly the same.
> 
> It does not iterate over the edges/faces. It computes directly with the geometry. This is because @RiccardoRossi says that the discussion from #4221 and #4222 will take long

OK, I didn't realize. Maybe we should rename it to `GeometryIntersection` to be more self-explanatory.
This new one is intended to be used when the local size of the intersected and intersecting geometry coincide, am I wrong?

rubenzorrilla commented:
> > > Yes, is this true, is the only reason I did it. If we move the flag from elements and conditions to geometrical object it can be avoided.
> > 
> > 
> > And are you sure that this doesn't give any problem? @pooyan-dadvand
> 
> No, and is something I should not decide, because is a design thing

Let's wait for @pooyan-dadvand opinion then.

pooyan-dadvand commented:
I am in favour of moving flags to the GeometricalEntity

loumalouomega commented:
> I am in favour of moving flags to the GeometricalEntity

OK, thanks this will simplify significantly the implementations (and will allow me to reduce copy paste the code)

pooyan-dadvand commented:
Please put the flags change in another PR so we can start discuss it in other thread without mixing up with these changes

loumalouomega commented:
> Please put the flags change in another PR so we can start discuss it in other thread without mixing up with these changes

Yes, I was planning to cherry-pick and create an independent PR

loumalouomega committed:
[Core] Moving flag from Condition/Element to GeometricalObject

Basically that. The flag is common between the geometries and conditions, but currently both derive from Flag instaed of bein included on the GeometricalObject

loumalouomega committed:
Reducing overhead of Element/Condition

loumalouomega committed:
Merge branch 'master' into core/FindIntersectedGeometricalObjectsProcess-entities

loumalouomega commented:
@rubenzorrilla I included #4746 and reduced overhead as you suggested

loumalouomega committed:
Serializer correction

loumalouomega committed:
Exposing Flags to python

loumalouomega committed:
@jcotela commentaries + doxygen

loumalouomega committed:
Merge branch 'master' into core/FindIntersectedGeometricalObjectsProcess-entities

loumalouomega committed:
Merge branch 'master' into core/FindIntersectedGeometricalObjectsProcess-entities

loumalouomega commented:
@rubenzorrilla @pooyan-dadvand  #4746 is merged, this is ready

loumalouomega commented:
Ping @rubenzorrilla

loumalouomega commented:
I have been speaking with @rubenzorrilla. The current implementation iterates over elements and conditions, as now the GeometricalObject uses flags some duplicated methods could be removed. In any case right now iterates over both entities, @rubenzorrilla doesn't like that, but adding some flags will increase complexity on my opinion, I can added anyway and then it will depend on the user/developer

rubenzorrilla commented:
Apart of the overhead, which can be significant in some applications, I am concerned about the Octree behaviour when it is created with both conditions and elements. According to @loumalouomega this shouldn't affect the results, since the let's say entities of interest can be now flagged.

My question is who should flag those entities of interest? Should them be flagged from outside the `FindIntersectedGeometricalObjectsProcess`? Can you clarify this @loumalouomega .

loumalouomega commented:
> Apart of the overhead, which can be significant in some applications, I am concerned about the Octree behaviour when it is created with both conditions and elements. According to @loumalouomega this shouldn't affect the results, since the let's say entities of interest can be now flagged.
> 
> My question is who should flag those entities of interest? Should them be flagged from outside the `FindIntersectedGeometricalObjectsProcess`? Can you clarify this @loumalouomega .

The entities are flagged on the FindIntersectedGeometricalObjectsProcess, even on the current master version

rubenzorrilla commented:
> > Apart of the overhead, which can be significant in some applications, I am concerned about the Octree behaviour when it is created with both conditions and elements. According to @loumalouomega this shouldn't affect the results, since the let's say entities of interest can be now flagged.
> > My question is who should flag those entities of interest? Should them be flagged from outside the `FindIntersectedGeometricalObjectsProcess`? Can you clarify this @loumalouomega .
> 
> The entities are flagged on the FindIntersectedGeometricalObjectsProcess, even on the current master version

My question is how the process knows if it should intersect conditions or elements against conditions or elements. With your enhancement, there are 4 possible options:

- condition - condition
- condition - element
- element - condition
- element - element

How the user controls which one is happening?

loumalouomega commented:
> > > Apart of the overhead, which can be significant in some applications, I am concerned about the Octree behaviour when it is created with both conditions and elements. According to @loumalouomega this shouldn't affect the results, since the let's say entities of interest can be now flagged.
> > > My question is who should flag those entities of interest? Should them be flagged from outside the `FindIntersectedGeometricalObjectsProcess`? Can you clarify this @loumalouomega .
> > 
> > 
> > The entities are flagged on the FindIntersectedGeometricalObjectsProcess, even on the current master version
> 
> My question is how the process knows if it should intersect conditions or elements against conditions or elements. With your enhancement, there are 4 possible options:
> 
>     * condition - condition
> 
>     * condition - element
> 
>     * element - condition
> 
>     * element - element
> 
> 
> How the user controls which one is happening?

Right now it intersects everything with everything, so you don't need to worry about that, before removing the templates it was as you say. The only things it affects is the flag, nothing else, if you know what are you looking for you don't need to be concerned, as long as you only focus your attention in the entities you are interested for. I will add the flags so you can have more control, but by default it will check everything

loumalouomega committed:
Merge branch 'master' into core/FindIntersectedGeometricalObjectsProcess-entities

loumalouomega committed:
Adding local flags

loumalouomega commented:
@rubenzorrilla I added the local flags

rubenzorrilla commented:
The point is that if now the user of this process needs to focus the attention on some specific intersections, this imply to change the derived processes (e.g. distance).

rubenzorrilla commented:
> The point is that if now the user of this process needs to focus the attention on some specific intersections, this imply to change the derived processes (e.g. distance).

I guess that this can know be handled with such local flags. In any case let's wait for @pooyan-dadvand  's opinion.

pooyan-dadvand commented:
I have two different concerns:
1. You had changed considerably the FindIntersetedGeometricalObject since starting this PR. Many of those changes where not necessary. Is it possible to make a clean reproduction of this PR from the current master version just to see it with minimum changes?
2. The current geometries of Kratos has limited intersection capability. So I'm not sure if all cases between volume/sking and volume/skin combinations would work in this scheme. Do you have tests for all configurations?

loumalouomega commented:
> I have two different concerns:
> 
>     1. You had changed considerably the FindIntersetedGeometricalObject since starting this PR. Many of those changes where not necessary. Is it possible to make a clean reproduction of this PR from the current master version just to see it with minimum changes?
Right now this PR just changes 3 files, and the PR have been created recently almost from scratch, copying the code from a different PR.  Not sure how to do it
> 
>     2. The current geometries of Kratos has limited intersection capability. So I'm not sure if all cases between volume/sking and volume/skin combinations would work in this scheme. Do you have tests for all configurations?

I have tests for the most common combinations, but anyway in case the intersection is not implemented the base geometry class will throw an error.

pooyan-dadvand reviewed (approved):
I am approving as the "new" code seems to be correct. Nevertheless, I would like to note that changing the code so drastically is far from any refactoring standard and please consider to avoid it in future. 

Please consider that if making a clean PR is difficult for you reviewing it is more complicated for the rest of the develoopers.

loumalouomega commented:
> I am approving as the "new" code seems to be correct. Nevertheless, I would like to note that changing the code so drastically is far from any refactoring standard and please consider to avoid it in future.
> 
> Please consider that if making a clean PR is difficult for you reviewing it is more complicated for the rest of the develoopers.

Yes, I understand, also the fact that there are 2 more PR pending of this one make it difficult to change something without breaking the other 2

rubenzorrilla commented:
Removing my blocking review.

loumalouomega commented:
Hurray
